File: Python3/2661_first_completely_painted_row_or_column.py
# https://leetcode.cn/problems/first-completely-painted-row-or-column/

class Solution:
    def firstCompleteIndex(self, arr, mat):
        """
        Intputs:
            arr: list[int]
            mat: list[list[int]]
        Outputs:
            res: int
        """
        m, n = len(mat), len(mat[0])
        val_idx = dict()
        for i in range(m):
            for j in range(n):
                val_idx[mat[i][j]] = (i, j)
        paintrow = [0 for _ in range(m)]
        paintcol = [0 for _ in range(n)]
        for i, val in enumerate(arr):
            row, col = val_idx[val]
            paintrow[row] += 1
            paintcol[col] += 1
            if paintrow[row] == n or paintcol[col] == m:
                return i
        return -1

File: Cpp/2661_first_completely_painted_row_or_column.cpp
// https://leetcode.cn/problems/first-completely-painted-row-or-column/

class Solution {
public:
    int firstCompleteIndex(vector<int>& arr, vector<vector<int>>& mat) {
        int m = mat.size(), n = mat[0].size();
        unordered_map<int, pair<int,int>> val_idx;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                val_idx[mat[i][j]] = make_pair(i, j);
            }
        }
        int res = -1;
        vector<int> paintrow(m, 0), paintcol(n, 0);
        for (int i = 0; i < arr.size(); ++i) {
            int val = arr[i];
            int row = val_idx[val].first, col = val_idx[val].second;
            paintrow[row] += 1;
            paintcol[col] += 1;
            if (paintrow[row] == n || paintcol[col] == m) {
                res = i;
                break;
            }
        }
        return res;
    }
};